libraries/Filter/HarmonicNotchFilter.cpp

```diff
@@ -212,46 +235,92 @@ void HarmonicNotchFilter<T>::expand_filter_count(uint16_t total_notches)
+    float A = _A;
+
+    // on some vehicles we want to treat zero or very low frequency as
+    // the minimum frequency, on others we want to disable the
+    // notch for low frequencies
+    const bool treat_low_freq_as_min = params->hasOption(HarmonicNotchFilterParams::Options::TreatLowAsMin);
+
+    if (treat_low_freq_as_min) {
+        /*
+          when we are treating low as min we don't want to collapse
+          the harmonics down to low frequencies
+         */
+        harmonic_min_freq *= harmonic_mul;
+    } else {
+        /*
+          disable if the notch is less than 25% of the min frequency
+        */
+        const float disable_freq = harmonic_min_freq * NOTCHFILTER_ATTENUATION_CUTOFF;
+        if (notch_center < disable_freq) {
+            notch.disable();
+            return;
+        }
 
-    _num_enabled_filters = 0;
-    // update all of the filters using the new center frequency and existing A & Q
-    for (uint8_t i = 0; i < HNF_MAX_HARMONICS && _num_enabled_filters < _num_filters; i++) {
-        if ((1U<<i) & _harmonics) {
-            const float notch_center = center_freq_hz * (i+1);
-            if (_composite_notches != 2) {
-                // only enable the filter if its center frequency is below the nyquist frequency
-                if (notch_center < nyquist_limit) {
-                    _filters[_num_enabled_filters++].init_with_A_and_Q(_sample_freq_hz, notch_center, _A, _Q);
-                }
-            }
-            if (_composite_notches > 1) {
-                float notch_center_double;
-                // only enable the filter if its center frequency is below the nyquist frequency
-                notch_center_double = notch_center * (1.0 - _notch_spread);
-                if (notch_center_double < nyquist_limit) {
-                    _filters[_num_enabled_filters++].init_with_A_and_Q(_sample_freq_hz, notch_center_double, _A, _Q);
-                }
-                // only enable the filter if its center frequency is below the nyquist frequency
-                notch_center_double = notch_center * (1.0 + _notch_spread);
-                if (notch_center_double < nyquist_limit) {
-                    _filters[_num_enabled_filters++].init_with_A_and_Q(_sample_freq_hz, notch_center_double, _A, _Q);
-                }
-            }
+        if (notch_center < harmonic_min_freq) {
+            /*
+              scale the attenuation so that we fade out the notch as
+              we get further below the min frequency. The attenuation
+              factor A goes to 1.0 (which means no attenuation)
+              Scaling the attenuation in this way means we don't get a
+              glitch at the disable point
+             */
+            A = linear_interpolate(A, 1.0, notch_center, harmonic_min_freq, disable_freq);
         }
     }
+
+    // don't let the notch go below the min frequency
+    notch_center = MAX(notch_center, harmonic_min_freq);
```

Comment:
Because the constrain is now happening here the the `FTN`/`FTNS` logs are no longer accurate. The now log the frequency we asked the notch to be at rather than where it is actually.

Now its different for each harmonic its a harder to get the logging right we certainly don't want to log harmonics separately. We could add a uint16 bitmask for enabled/disable with one bit per harmonic. You could then see the low harmonics disable as the center tracks down. But I guess it would be one bitmask per notch, so 12 more values.

I'm not sure.

Reply:
I had not spotted this, but I agree that it would be better to log the actual value being used - particularly below the min. Counter to this the current slewing code does not log the actual value, which is a pain.

Reply:
what about a FTST for "filter status" message with two bitmasks, one of enable status one or clamp status?